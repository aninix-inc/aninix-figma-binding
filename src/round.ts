/**
 * @description optimized version of rounding
 */
export const round = (number: number, fixed = 4): number => {
  if (fixed === 0) {
    // Use bitwise OR to perform a fast flooring operation
    return (number + 0.5) | 0
  }

  const multiplier = 10 ** fixed
  const rounded = (number * multiplier + 0.5) | 0
  return rounded / multiplier
}
