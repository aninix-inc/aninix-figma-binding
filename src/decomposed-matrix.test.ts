import { describe, test, expect } from 'bun:test'
import {
  decomposedMatrix,
  type DecomposedMatrix,
  unflatten,
} from './decomposed-matrix'
import { round } from './round'

export type ComponentMatrices = {
  translationMatrix: number[]
  rotationMatrix: number[]
  scalingMatrix: number[]
  skewingMatrix: number[]
}

const multiplyMatrices = (...matrices: number[][]): number[] => {
  return matrices.reduce(
    (result, matrix) => {
      const [a, b, c, d, e, f] = result
      const [g, h, i, j, k, l] = matrix

      result[0] = a * g + b * i
      result[1] = a * h + b * j
      result[2] = c * g + d * i
      result[3] = c * h + d * j
      result[4] = e * g + f * i + k
      result[5] = e * h + f * j + l

      return result
    },
    [1, 0, 0, 1, 0, 0]
  )
}

const composeMatrix = (components: DecomposedMatrix): number[] => {
  const { translation, rotation, scaling, skewing } = components

  const rads = Math.PI / 180
  const translationMatrix = [1, 0, 0, 1, translation.x, translation.y]
  const rotationMatrix = [
    Math.cos(rotation.x * rads),
    -Math.sin(rotation.x * rads),
    Math.sin(rotation.x * rads),
    Math.cos(rotation.x * rads),
    0,
    0,
  ]
  const scalingMatrix = [scaling.x, 0, 0, scaling.y, 0, 0]
  const skewingMatrix = [
    1,
    Math.tan(skewing.y * rads),
    Math.tan(skewing.x * rads),
    1,
    0,
    0,
  ]

  return multiplyMatrices(
    rotationMatrix,
    skewingMatrix,
    scalingMatrix,
    translationMatrix
  )
}

const testMatrix = [
  [0.87, 0.87, 72.99],
  [0.5, -0.5, 62.38],
] as [[number, number, number], [number, number, number]]

describe('decomposedMatrix', () => {
  test('decomposedMatrix', () => {
    const components = decomposedMatrix(testMatrix)
    const result = composeMatrix(components).map((i) => round(i, { fixed: 2 }))
    expect(unflatten(result as any)).toMatchSnapshot()
    expect(result).toMatchSnapshot()
  })
})
