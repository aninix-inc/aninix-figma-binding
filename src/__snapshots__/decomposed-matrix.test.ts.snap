// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`decomposedMatrix decomposedMatrix 1`] = `
[
  [
    NaN,
    NaN,
    NaN,
  ],
  [
    NaN,
    NaN,
    NaN,
  ],
]
`;

exports[`decomposedMatrix decomposedMatrix 2`] = `
[
  NaN,
  NaN,
  NaN,
  NaN,
  NaN,
  NaN,
]
`;
